### PR TITLE
KEYCLOAK-18824 Simplify MapStorageTransaction and move registerEntityForChanges to CHM transaction

### PR DESCRIPTION
**Pull request by hmlnarik**
This PR contains #8272 to prevent merging conflicts

### CODE REVIEW

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 1]
```diff
@@ -535,7 +526,7 @@ public UserModel getUserByEmail(RealmModel realm, String email) {
             throw new ModelDuplicateException("Multiple users with email '" + email + "' exist in Keycloak.");
         }
 
-        MapUserEntity<K> userEntity = registerEntityForChanges(tx, usersWithEmail.get(0));
+        MapUserEntity<K> userEntity = usersWithEmail.get(0);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:529]: Please refactor this method to use `entityToAdapterFunc` on line 539.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 2]
```diff
@@ -336,7 +333,7 @@ public UserModel addUser(RealmModel realm, String id, String username, boolean a
         entity.setUsername(username.toLowerCase());
         entity.setCreatedTimestamp(Time.currentTimeMillis());
 
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java]: It seems that on this line we had `entity =` after the first commit and then the second commit removed it. Should `entity =` remain there?

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapPermissionTicketStore.java [hunk 3]
```diff
@@ -128,7 +127,7 @@ public PermissionTicket create(String resourceId, String scopeId, String request
         entity.setOwner(owner);
         entity.setResourceServerId(resourceServer.getId());
 
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapPermissionTicketStore.java]: the same

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java [hunk 4]
```diff
@@ -96,7 +95,7 @@ public Policy create(AbstractPolicyRepresentation representation, ResourceServer
         entity.setName(representation.getName());
         entity.setResourceServerId(resourceServer.getId());
         
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapPolicyStore.java]: the same

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapResourceServerStore.java [hunk 5]
```diff
@@ -82,7 +81,7 @@ public ResourceServer create(String clientId) {
 
         MapResourceServerEntity<K> entity = new MapResourceServerEntity<>(resourceServerStore.getKeyConvertor().fromString(clientId));
 
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapResourceServerStore.java]: the same

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapResourceStore.java [hunk 6]
```diff
@@ -96,7 +95,7 @@ public Resource create(String id, String name, ResourceServer resourceServer, St
         entity.setResourceServerId(resourceServer.getId());
         entity.setOwner(owner);
 
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapResourceStore.java]: the same

File: model/map/src/main/java/org/keycloak/models/map/authorization/MapScopeStore.java [hunk 7]
```diff
@@ -94,7 +93,7 @@ public Scope create(String id, String name, ResourceServer resourceServer) {
         entity.setName(name);
         entity.setResourceServerId(resourceServer.getId());
 
-        entity = tx.create(entity);
+        tx.create(entity);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/authorization/MapScopeStore.java]: the same

**Review by mhajas — COMMENTED** [review]
Thank you for the changes @hmlnarik. The idea of adding registering entity into the transaction totally makes sense! We probably won't be able to have unregistered entity from the JPA anyway, so it doesn't make sense to have this concept in the providers.

I looked much closer on users, groups and authorization than on other areas. I assumed @vramik and @martin-kanis to look at others closer.

I added few comments into the code, but in general it looks good.

**Comment by hmlnarik** [comment]
@mhajas Thanks for the review. Could you please re-review?

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java [hunk 8]
```diff
@@ -98,16 +100,40 @@ protected void addTask(K key, MapTaskWithValue task) {
         tasks.merge(taskKey, task, MapTaskCompose::new);
     }
 
+    /**
+     * Returns a deep clone of an entity. If the clone is already in the transaction, returns this one.
+     * <p>
+     * Usually used before giving an entity from a source back to the caller,
+     * to prevent changing it directly in the data store, but to keep transactional properties.
+     * @param <K>
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java:108]: Parameter K is not used here.

**Review by vramik — COMMENTED** [review]
Thank you @hmlnarik for the PR, it looks great. I've just two comments regarding javadoc.

**Review by mhajas — DISMISSED** [review]
Thanks for the changes @hmlnarik. Approved

**Review by martin-kanis — CHANGES_REQUESTED** [review]

**Review by vramik — APPROVED** [review]
Thank you @hmlnarik

**Review by martin-kanis — APPROVED** [review]
Thanks @hmlnarik

**Pull request merged**